server/src/main/java/org/candlepin/controller/ProductManager.java

```diff
@@ -84,121 +74,255 @@ public ProductManager(EntitlementCertificateGenerator entitlementCertGenerator,
+        }
+
+        Collection<? extends ProductInfo> provided = pinfo.getProvidedProducts();
+        if (provided != null) {
+            for (ProductInfo pp : provided) {
+                if (pp == null) {
+                    // TODO: Make this a custom exception. MalformedChildReferenceException, perhaps?
+                    throw new IllegalArgumentException("product contains null provided product reference");
+                }
+
+                if (pp.getId() == null || pp.getId().isEmpty()) {
+                    // TODO: Make this a custom exception. MalformedChildReferenceException, perhaps?
+                    throw new IllegalArgumentException("product references a product that lacks a valid ID");
+                }
+
+                pids.add(pp.getId());
+            }
+        }
+
+        if (!pids.isEmpty()) {
+            output = this.ownerProductCurator.getProductsByIds(owner, pids).list().stream()
+                .collect(Collectors.toMap(Product::getId, Function.identity()));
+
+            pids.removeAll(output.keySet());
+            if (!pids.isEmpty()) {
+                // TODO: Make this a custom exception. MalformedChildReferenceException, perhaps?
+                throw new IllegalArgumentException(
+                    "product references one or more non-existent products: " + pids);
+            }
+        }
+        else {
+            output = new HashMap<>();
+        }
+
+        return output;
+    }
+
+    /**
+     * Resolves all of the products referenced by the given product info, returning a map that
+     * contains all resolved product entities mapped by product ID. If a product reference cannot
+     * be resolved, this method throws an exception.
+     *
+     * @param pinfo
+     *  the product info for which to resolve product references
      *
      * @throws IllegalArgumentException
-     *  if dto is null or incomplete, or owner is null
+     *  if the product provided references a product which cannot be resolved
      *
      * @return
-     *  a new Product instance representing the specified product for the given owner
+     *  a map containing all of the resolved products referenced, mapped by product ID
      */
-    public Product createProduct(ProductDTO dto, Owner owner) {
-        if (dto == null) {
-            throw new IllegalArgumentException("dto is null");
+    private Map<String, Content> resolveContentRefs(Owner owner, ProductInfo pinfo) {
+        Set<String> cids = new HashSet<>();
+        Map<String, Content> output;
+
+        Collection<? extends ProductContentInfo> productContent = pinfo.getProductContent();
+        if (productContent != null) {
+            for (ProductContentInfo pcinfo : productContent) {
+                ContentInfo cinfo = pcinfo != null ? pcinfo.getContent() : null;
+
+                if (cinfo == null) {
+                    // TODO: Make this a custom exception. MalformedChildReferenceException, perhaps?
```

Comment:
I didn't touch the ISEs since I feel that's still a decent enough fit, but the TODOs have been addressed with the new ```MalformedEntityReferenceException```.